File: hdar/cpp/src/apple_container.cpp
#include "hdar/apple_container.hpp"
#include <cstdio>
#include <array>
#include <chrono>
#include <filesystem>

namespace fs = std::filesystem;

namespace hdar {

AppleContainerProvider::AppleContainerProvider() {}

std::string AppleContainerProvider::exec_cli(const std::string& cmd, int* exit_code) {
    std::string result;
    FILE* pipe = popen(cmd.c_str(), "r");
    if (!pipe) {
        if (exit_code) *exit_code = -1;
        return "";
    }
    std::array<char, 4096> buf;
    while (fgets(buf.data(), buf.size(), pipe) != nullptr)
        result += buf.data();
    int status = pclose(pipe);
    if (exit_code) *exit_code = WEXITSTATUS(status);
    return result;
}

bool AppleContainerProvider::is_available() {
    int rc = -1;
    exec_cli("which container 2>/dev/null", &rc);
    return rc == 0;
}

RuntimeRecord AppleContainerProvider::materialize(
    const std::string& runtime_id,
    const std::string& workspace_path,
    const std::string& image,
    const std::string& cpu_limit,
    const std::string& memory_limit,
    const std::string& network_policy) {

    // 1. Create the container (stopped)
    std::string img = image.empty() ? "ubuntu:24.04" : image;
    std::string create_cmd = "container create --name " + runtime_id;
    if (!cpu_limit.empty()) create_cmd += " -c " + cpu_limit;
    if (!memory_limit.empty()) create_cmd += " -m " + memory_limit;
    create_cmd += " " + img + " sleep infinity 2>&1";

    int rc = -1;
    std::string create_out = exec_cli(create_cmd, &rc);

    RuntimeRecord rec;
    rec.provider = "apple-container";
    rec.runtime_id = runtime_id;
    rec.image_digest = img;
    rec.cpu_limit = cpu_limit;
    rec.memory_limit = memory_limit;
    rec.workspace_mount = workspace_path;
    rec.network_policy = network_policy;
    rec.start_timestamp = epoch_seconds();

    if (rc != 0) {
        rec.exists = false;
        runtimes_[runtime_id] = rec;
        return rec;
    }

    // 2. Start the container
    std::string start_cmd = "container start " + runtime_id + " 2>&1";
    rc = -1;
    exec_cli(start_cmd, &rc);

    rec.exists = (rc == 0);

    // 3. Inspect to get VM identity
    if (rec.exists) {
        std::string inspect_cmd = "container inspect " + runtime_id + " 2>/dev/null";
        std::string inspect_out = exec_cli(inspect_cmd);
        // Try to parse JSON inspect output for VM details
        try {
            JsonValue inspect_json = parse_json(inspect_out);
            if (inspect_json.type == JsonValue::Type::Array && !inspect_json.array_val.empty()) {
                auto& cfg = inspect_json.array_val[0].get("configuration");
                rec.vm_identity = cfg.get("id").string_val;
            }
        } catch (...) {
            rec.vm_identity = inspect_out.substr(0, 256);
        }
    }

    runtimes_[runtime_id] = rec;
    return rec;
}

ExecutionResult AppleContainerProvider::execute(
    const std::string& runtime_id,
    const std::string& operation_type,
    const std::string& command,
    int timeout) {

    ExecutionResult result;
    result.operation_type = operation_type;
    result.command = command;

    auto it = runtimes_.find(runtime_id);
    if (it == runtimes_.end() || !it->second.exists) {
        result.stderr_text = "runtime not found or not running";
        return result;
    }

    // container exec <runtime_id> sh -c "<command>"
    std::string cmd = "container exec " + runtime_id + " sh -c \"" + command + "\" 2>&1";

    auto start = std::chrono::steady_clock::now();
    int rc = -1;
    result.stdout_text = exec_cli(cmd, &rc);
    result.exit_code = rc;
    result.success = (rc == 0);

    auto end = std::chrono::steady_clock::now();
    result.duration_ms = std::chrono::duration<double, std::milli>(end - start).count();

    return result;
}

RuntimeRecord AppleContainerProvider::stop(const std::string& runtime_id) {
    auto it = runtimes_.find(runtime_id);
    if (it == runtimes_.end())
        throw std::runtime_error("runtime not found: " + runtime_id);

    exec_cli("container kill " + runtime_id + " 2>/dev/null");
    it->second.stop_timestamp = epoch_seconds();
    return it->second;
}

RuntimeRecord AppleContainerProvider::destroy(const std::string& runtime_id) {
    auto it = runtimes_.find(runtime_id);
    if (it == runtimes_.end())
        throw std::runtime_error("runtime not found: " + runtime_id);

    // Kill if still running, then force delete
    exec_cli("container kill " + runtime_id + " 2>/dev/null");
    int rc = -1;
    std::string del_output = exec_cli("container delete --force " + runtime_id + " 2>&1", &rc);

    it->second.delete_timestamp = epoch_seconds();
    it->second.exists = false;

    // Post-delete inspection — verify container is gone
    int inspect_rc = -1;
    std::string inspect_out = exec_cli("container inspect " + runtime_id + " 2>&1", &inspect_rc);

    JsonValue inspection = JsonValue::object();
    inspection["exists"] = JsonValue::boolean(inspect_rc == 0);
    inspection["provider"] = JsonValue::string("apple-container");
    inspection["runtime_id"] = JsonValue::string(runtime_id);
    inspection["delete_exit_code"] = JsonValue::integer(rc);
    inspection["inspect_exit_code"] = JsonValue::integer(inspect_rc);
    inspection["inspect_output"] = JsonValue::string(inspect_out);
    it->second.post_delete_inspection = inspection;

    return it->second;
}

JsonValue AppleContainerProvider::inspect(const std::string& runtime_id) {
    JsonValue v = JsonValue::object();

    // Check if container exists via CLI
    // container inspect returns non-zero if container doesn't exist
    int rc = -1;
    std::string inspect_out = exec_cli("container inspect " + runtime_id + " 2>/dev/null", &rc);

    bool exists = (rc == 0);

    v["exists"] = JsonValue::boolean(exists);
    v["provider"] = JsonValue::string("apple-container");
    v["runtime_id"] = JsonValue::string(runtime_id);

    if (exists) {
        // Parse inspect output for state
        try {
            JsonValue parsed = parse_json(inspect_out);
            if (parsed.type == JsonValue::Type::Array && !parsed.array_val.empty()) {
                auto& item = parsed.array_val[0];
                v["state"] = item.get("status").get("state");
                v["platform"] = item.get("configuration").get("platform");
            }
        } catch (...) {
            v["inspect_raw"] = JsonValue::string(inspect_out);
        }
    }

    auto it = runtimes_.find(runtime_id);
    if (it != runtimes_.end()) {
        v["start_timestamp"] = JsonValue::number(it->second.start_timestamp);
        if (it->second.stop_timestamp)
            v["stop_timestamp"] = JsonValue::number(*it->second.stop_timestamp);
        if (it->second.delete_timestamp)
            v["delete_timestamp"] = JsonValue::number(*it->second.delete_timestamp);
    }

    return v;
}

std::vector<std::string> AppleContainerProvider::list_runtimes() {
    std::vector<std::string> result;
    std::string output = exec_cli("container list --all --quiet 2>/dev/null");
    std::istringstream iss(output);
    std::string line;
    while (std::getline(iss, line)) {
        while (!line.empty() && (line.back() == '\r' || line.back() == '\n'))
            line.pop_back();
        if (!line.empty())
            result.push_back(line);
    }
    return result;
}

} // namespace hdar

File: hdar/hdar_core/providers/apple_container.py
"""Apple container provider — real isolated Linux VM execution.

Uses Apple's `container` CLI (https://github.com/apple/container) to
create lightweight VM-backed Linux containers on Apple silicon.

Each container runs in its own lightweight VM with explicit CPU and
memory limits. This provider records real runtime identity, resource
allocation, and post-destruction absence proof.

Requires:
    brew install container
    container system kernel set --recommended
    container image pull ubuntu:24.04

The provider will raise RuntimeError if the CLI is not available.
"""

from __future__ import annotations

import json
import shutil
import subprocess
import time
from typing import Optional

from .base import ProviderBase, RuntimeRecord, ExecutionResult


class AppleContainerProvider(ProviderBase):
    """Real Apple container provider — VM-backed Linux isolation.

    Each container is a lightweight VM running Linux on Apple silicon.
    The provider records:
      - Real VM identity (IP, OS, arch from inspect)
      - Resource allocation (CPU, memory)
      - Post-destruction absence proof (inspect returns not-found)
    """

    def __init__(self):
        self._cli = self._find_cli()
        self._runtimes: dict[str, RuntimeRecord] = {}

    def _find_cli(self) -> str:
        path = shutil.which("container")
        if not path:
            raise RuntimeError(
                "Apple 'container' CLI not found. Install with: brew install container"
            )
        return path

    @property
    def name(self) -> str:
        return "apple-container"

    def _run_cli(self, args: list, timeout: int = 120) -> subprocess.CompletedProcess:
        return subprocess.run(
            [self._cli] + args,
            capture_output=True,
            text=True,
            timeout=timeout,
        )

    def materialize(
        self,
        runtime_id: str,
        workspace_path: str,
        image: str = "ubuntu:24.04",
        cpu_limit: str = "2",
        memory_limit: str = "512m",
        network_policy: str = "default",
    ) -> RuntimeRecord:
        """Create and start a real VM-backed container.

        Uses `container run -d` with `sleep infinity` as init process
        to keep the VM alive for exec commands.
        """
        run_args = [
            "run",
            "--name", runtime_id,
            "-c", str(cpu_limit),
            "-m", memory_limit,
            "-d",
        ]

        if workspace_path:
            run_args.extend(["-v", f"{workspace_path}:/workspace"])

        if network_policy == "none":
            run_args.extend(["--network", "none"])

        run_args.extend([image, "sleep", "infinity"])

        result = self._run_cli(run_args, timeout=60)
        if result.returncode != 0:
            raise RuntimeError(
                f"container run failed: {result.stderr.strip()}"
            )

        info = self.inspect(runtime_id)

        record = RuntimeRecord(
            provider=self.name,
            runtime_id=runtime_id,
            image_digest=image,
            vm_identity=info.get("id", runtime_id),
            cpu_limit=str(cpu_limit),
            memory_limit=memory_limit,
            workspace_mount=workspace_path,
            network_policy=network_policy,
            start_timestamp=time.time(),
            exists=True,
        )
        self._runtimes[runtime_id] = record
        return record

    def execute(
        self,
        runtime_id: str,
        operation_type: str,
        command: str,
        timeout: int = 60,
    ) -> ExecutionResult:
        """Execute a command inside the running container."""
        if runtime_id not in self._runtimes:
            return ExecutionResult(
                operation_type=operation_type,
                command=command,
                exit_code=-1,
                stderr="runtime not found",
                success=False,
            )

        start = time.time()
        result = self._run_cli(
            ["exec", runtime_id, "sh", "-c", command],
            timeout=timeout,
        )
        duration = (time.time() - start) * 1000

        return ExecutionResult(
            operation_type=operation_type,
            command=command,
            exit_code=result.returncode,
            stdout=result.stdout,
            stderr=result.stderr,
            duration_ms=duration,
            success=result.returncode == 0,
        )

    def stop(self, runtime_id: str) -> RuntimeRecord:
        if runtime_id not in self._runtimes:
            return RuntimeRecord(
                provider=self.name, runtime_id=runtime_id, exists=False
            )
        self._run_cli(["stop", runtime_id])
        record = self._runtimes[runtime_id]
        record.stop_timestamp = time.time()
        return record

    def destroy(self, runtime_id: str) -> RuntimeRecord:
        """Destroy the container and record absence proof."""
        if runtime_id not in self._runtimes:
            return RuntimeRecord(
                provider=self.name, runtime_id=runtime_id, exists=False
            )
        record = self._runtimes[runtime_id]

        if record.stop_timestamp is None:
            self.stop(runtime_id)

        self._run_cli(["rm", runtime_id])
        record.delete_timestamp = time.time()
        record.exists = False
        record.post_delete_inspection = self.inspect(runtime_id)
        del self._runtimes[runtime_id]
        return record

    def inspect(self, runtime_id: str) -> dict:
        """Query provider for runtime state. Returns {"exists": False} if gone."""
        result = self._run_cli(["inspect", runtime_id])
        if result.returncode != 0:
            return {
                "exists": False,
                "provider": self.name,
                "runtime_id": runtime_id,
                "error": result.stderr.strip(),
            }
        try:
            data = json.loads(result.stdout)
            if isinstance(data, list) and len(data) > 0:
                entry = data[0]
                return {
                    "exists": True,
                    "id": entry.get("id", runtime_id),
                    "state": entry.get("status", {}).get("state", "unknown"),
                    "image": entry.get("configuration", {}).get("image", {}).get("reference", ""),
                    "os": entry.get("configuration", {}).get("platform", {}).get("os", ""),
                    "arch": entry.get("configuration", {}).get("platform", {}).get("architecture", ""),
                    "cpus": entry.get("configuration", {}).get("resources", {}).get("cpus", 0),
                    "memory": entry.get("configuration", {}).get("resources", {}).get("memoryInBytes", 0),
                    "started": entry.get("status", {}).get("startedDate", ""),
                    "raw": entry,
                }
            return {"exists": True, "raw": data}
        except json.JSONDecodeError:
            return {
                "exists": True,
                "provider": self.name,
                "runtime_id": runtime_id,
                "raw": result.stdout,
            }

    def list_runtimes(self) -> list:
        """List all containers (running and stopped)."""
        result = self._run_cli(["ls", "-a"])
        if result.returncode != 0:
            return []
        lines = result.stdout.strip().split("\n")
        if len(lines) < 2:
            return []
        ids = []
        for line in lines[1:]:
            parts = line.split()
            if parts:
                ids.append(parts[0])
        return ids

    def verify_destruction(self, runtime_id: str) -> bool:
        """Verify runtime no longer exists — the absence proof."""
        listing = self.list_runtimes()
        if runtime_id in listing:
            return False
        inspection = self.inspect(runtime_id)
        return not inspection.get("exists", False)
